src/main/java/de/tum/in/www1/artemis/domain/Result.java

```diff
@@ -470,19 +470,37 @@ public void filterSensitiveInformation() {
     }
 
     /**
-     * Remove all feedbacks marked with visibility never.
+     * Removes all feedback details that should not be passed to the student.
      *
      * @param isBeforeDueDate if feedbacks marked with visibility 'after due date' should also be removed.
      */
     public void filterSensitiveFeedbacks(boolean isBeforeDueDate) {
+        filterSensitiveFeedbacks(isBeforeDueDate, participation.getExercise());
+    }
+
+    /**
+     * Removes all feedback details that should not be passed to the student.
+     *
+     * @param isBeforeDueDate if feedbacks marked with visibility 'after due date' should also be removed.
+     * @param exercise        the exercise this result corresponds to. Use to determine if test case names should be shown.
+     */
+    public void filterSensitiveFeedbacks(boolean isBeforeDueDate, Exercise exercise) {
         feedbacks.removeIf(Feedback::isInvisible);
 
         if (isBeforeDueDate) {
             feedbacks.removeIf(Feedback::isAfterDueDate);
         }
-        // TODO: this is not good code!
-        setTestCaseCount((int) feedbacks.stream().filter(Feedback::isTestFeedback).count());
-        setPassedTestCaseCount((int) feedbacks.stream().filter(Feedback::isTestFeedback).filter(feedback -> Boolean.TRUE.equals(feedback.isPositive())).count());
+
+        if (exercise instanceof ProgrammingExercise programmingExercise) {
+            var testCaseFeedback = feedbacks.stream().filter(Feedback::isTestFeedback).toList();
+            if (!programmingExercise.getShowTestNamesToStudents()) {
+                testCaseFeedback.forEach(feedback -> feedback.setText(null));
+            }
+
+            // TODO: this is not good code!
+            setTestCaseCount(testCaseFeedback.size());
+            setPassedTestCaseCount((int) testCaseFeedback.stream().filter(feedback -> Boolean.TRUE.equals(feedback.isPositive())).count());
```

Comment:
Could be moved to a private helper method `filterTestCaseFeedback(ProgrammingExercise)`, but also okay as-is.